File: realtime-macos/Sources/PAPAluLive/IdleAnimationPlan.swift
enum IdleSwayDirection: Equatable {
    case left
    case right
}

struct IdleSwayStep: Equatable {
    let horizontalOffset: Double
    let rotationDegrees: Double
    let duration: Double
    let holdDuration: Double
}

struct IdleAnimationPlan {
    let configuration: IdleMotionConfiguration

    init(configuration: IdleMotionConfiguration = .gentle) {
        self.configuration = configuration
    }

    func swayStep(
        direction: IdleSwayDirection,
        durationRandomUnit: Double,
        holdRandomUnit: Double
    ) -> IdleSwayStep {
        let sign = direction == .left ? -1.0 : 1.0
        return IdleSwayStep(
            horizontalOffset: sign * configuration.horizontalOffset,
            rotationDegrees: sign * configuration.rotationDegrees,
            duration: map(durationRandomUnit, into: configuration.durationRange),
            holdDuration: map(holdRandomUnit, into: configuration.holdRange)
        )
    }

    private func map(
        _ randomUnit: Double,
        into range: ClosedRange<Double>
    ) -> Double {
        let finiteUnit = randomUnit.isFinite ? randomUnit : 0
        let clampedUnit = min(1, max(0, finiteUnit))
        return range.lowerBound
            + (range.upperBound - range.lowerBound) * clampedUnit
    }
}

File: realtime-windows/PAPAluLive.Core/IdleAnimationPlan.cs
namespace PAPAluLive.Core;

public enum IdleSwayDirection
{
    Left,
    Right,
}

public sealed record IdleSwayStep(
    double HorizontalOffset,
    double RotationDegrees,
    double Duration,
    double HoldDuration);

public sealed class IdleAnimationPlan
{
    private readonly IdleMotionConfiguration configuration;

    public IdleAnimationPlan(IdleMotionConfiguration? configuration = null)
    {
        this.configuration = configuration ?? IdleMotionConfiguration.Gentle;
    }

    public IdleSwayStep GetStep(
        IdleSwayDirection direction,
        double durationRandomUnit,
        double holdRandomUnit)
    {
        var sign = direction == IdleSwayDirection.Left ? -1.0 : 1.0;
        return new IdleSwayStep(
            HorizontalOffset: sign * configuration.HorizontalOffset,
            RotationDegrees: sign * configuration.RotationDegrees,
            Duration: Map(
                durationRandomUnit,
                configuration.MinimumDuration,
                configuration.MaximumDuration),
            HoldDuration: Map(
                holdRandomUnit,
                configuration.MinimumHold,
                configuration.MaximumHold));
    }

    private static double Map(double randomUnit, double minimum, double maximum)
    {
        var finite = double.IsFinite(randomUnit) ? randomUnit : 0;
        var clamped = Math.Clamp(finite, 0, 1);
        return minimum + (maximum - minimum) * clamped;
    }
}
